text/2020-03-07-copr-chunk.md

```diff
@@ -0,0 +1,248 @@
+```
+
+We should modify the `VectorValue` to:
+
+```rust
+#[derive(Debug, PartialEq, Clone)]
+pub enum VectorValue {
+    Int(TypedVector<Int>),
+    Real(TypedVector<Real>),
+    Decimal(TypedVector<Decimal>),
+    Bytes(TypedVector<Bytes>),
+    DateTime(TypedVector<DateTime>),
+    Duration(TypedVector<Duration>),
+    Json(TypedVector<Json>),
+}
+```
+
+`TypedVector` implemented [chunk column format](https://pingcap.com/blog-cn/tidb-source-code-reading-10/).
+
+The original definition of `Bytes` is `type Bytes = Vec<u8>`, we have to modify it to `type Bytes<'a> = &'a [u8];` due to the memory of `Bytes` is actually owned by `TypedVector`.
+
+The `ScalarRef` is impossible to keep compatibility.
+
+```rust
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ScalarValueRef<'a> {
+    // Fixed sized.
+    Int(&'a Option<Int>),
+    // ...
+    Bytes(Option<&Bytes>),
+    Json(Option<&Json>),
+}
+```
+
+We have to modify it to:
+
+```rust
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ScalarValueRef<'a> {
+    // Fixed sized.
+    Int(Option<&'a Int>),
+    // ...
+    Bytes(Option<Bytes>),
+    Json(Option<JsonRef>),
+}
+```
+
+We also need to modify all rpn_function signature from
+
+```rust
+#[rpn_fn]
+pub fn logical_and(lhs: &Option<i64>, rhs: &Option<i64>) -> Result<Option<i64>>;
+
+#[rpn_fn]
+pub fn md5(arg: &Option<Bytes>) -> Result<Option<Vec<u8>>>;
+```
+
+to
+
+```rust
+#[rpn_fn]
+pub fn logical_and(lhs: Option<&i64>, rhs: Option<&i64>) -> Result<Option<i64>>;
+
+#[rpn_fn]
+pub fn md5(arg: Option<Bytes>) -> Result<Option<Vec<u8>>>;
+```
+
+The input has been optimized to `&[u8]`, but the result should keep `Vec<u8>`.
+
+Although we can implement all builtin functions based on the new ScalarRef args, it's not efficient enough. We have to read both bitmap and data map to construct an `Option<&T>`. Instead, for most builtin function, if some of the input args is NULL, then the output will be NULL. We can handle NULL by default.
+
+
+```rust
+#[rpn_fn]
+pub fn md5(arg: Bytes) -> Result<Option<Vec<u8>>>;
+```
+
+We should add new attribute `nullable` in `rpn_fn`, so that some special functions can handle NULL manually.
+
+```rust
+#[rpn_fn(nullable = true)]
+pub fn md5(arg: Option<Bytes>) -> Result<Option<Vec<u8>>>;
+```
+
+For most functions, `nullable` is not required.
+
+Here are still some overheads in the return value `Vec<u8>`, sometimes we have to allocate a new Vec and append it to `VectorValue` immediately, the allocation is not required. A typical case is `substr`, we should pass the `VectorValue` into every function. To hide the implementation defail, we can create a wrapper:
+
+```rust
+pub struct Guard {
+    _inner: (),
+}
+
+pub trait Writer<T> {
+    write(self, Option<T>) -> Guard;
+}
+
+Writer::<Bytes>;
+Writer::<JsonRef>;
+```
+
+Then the signature can be modified to:
+
+```rust
+#[rpn_fn]
+pub fn md5(arg: Bytes, writer: Writer<Bytes>) -> Result<Guard> {
+    writer.write(hex_digest(MessageDigest::md5(), arg));
```

Comment:
note that in this form, it is possible to first write something, then return an error.

Reply:
If Err is returned, we should call discard in framework